walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,97 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Set;
+import java.util.Collections;
+import java.util.Comparator;
 
 public final class FindMeetingQuery {
+
+  static final int START_OF_DAY = TimeRange.START_OF_DAY;
+  static final int END_OF_DAY = TimeRange.END_OF_DAY;
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> allAttendees = new ArrayList<>();
+    allAttendees.addAll(request.getAttendees());
+    allAttendees.addAll(request.getOptionalAttendees());
+    MeetingRequest meetingWithEveryone = new MeetingRequest(allAttendees, request.getDuration());
+
+    Collection<TimeRange> meetingTimesWithAll = findTime(events, meetingWithEveryone);
+
+    // Returns meeting times based on if there are available times with optional attendees 
+    if (meetingTimesWithAll.size() > 0 || request.getAttendees().isEmpty()) {
+      return meetingTimesWithAll;
+    } else {
+      return findTime(events, request);
+    }
+  }
+
+  public Collection<TimeRange> findTime(Collection<Event> events, MeetingRequest request) {
+    
+    List<TimeRange> possibleMeetingTimes = new ArrayList<>();
+    List<TimeRange> blockedTimes = new ArrayList<>();
+    int requestDuration = (int)request.getDuration();
+    Collection<String> requestAttendees = request.getAttendees();
+    int numEvents = events.size();
+    int availableStart = START_OF_DAY;
+
+    // No meeting times if duration is greater than whole day 
+    if (requestDuration > TimeRange.WHOLE_DAY.duration()) {
+      return possibleMeetingTimes;
+    }
+
+    // Check if no events happening
+    if (numEvents == 0) {
+      possibleMeetingTimes.add(TimeRange.WHOLE_DAY);
+      return possibleMeetingTimes;
+    }
+
+    // Find events in which people in meeting request are in the event 
+    for (Event event : events) {
+      if (checkOverlap(event.getAttendees(), requestAttendees)) {
+        blockedTimes.add(event.getWhen());
+      }
+    }
+
+    // Order blocked meetings times based on start time 
+    Collections.sort((List)blockedTimes, TimeRange.ORDER_BY_START);
+
+    for (TimeRange e : blockedTimes) {
+      int eventStart = e.start();
+      int eventEnd = e.end();
+
+      // Check for first event and overlapping events 
+      if (eventStart == START_OF_DAY || eventStart < availableStart && eventEnd > availableStart) {
+        availableStart = eventEnd;
+      }
+      
+      if (availableStart <= eventStart) {
+        if (hasEnoughMeetingTime(eventStart, availableStart, requestDuration)) {
+          possibleMeetingTimes.add(TimeRange.fromStartEnd(availableStart, eventStart, false));
```

Comment:
I know this is obnoxious, but I had to dig around to figure out what that boolean is doing. One simple trick we use often for passing a value into a function where the variable name doesn't explain the function parameter/usage:

```java
TimeRange.fromStartEnd(availableStart, eventStart, /* inclusive= */ false)
```

I'd recommend doing this everywhere you use this method.

Reply:
sounds good!! added